### PR TITLE
feat: Supports UUID column

### PR DESCRIPTION
**Pull request by huaxingao**
## Which issue does this PR close?

Closes #.

## Rationale for this change

Supports UUID column. This is for Iceberg/Comet integration 

## How are these changes tested?

This has been tested locally using iceberg

### CODE REVIEW

**Comment by huaxingao** [comment]
> should we get this covered by tests

This is for iceberg/comet integration. I don't think there is an easy way to test this now. I have tested this on my local, though.

**Comment by comphead** [comment]
> > should we get this covered by tests
> 
> This is for iceberg/comet integration. I don't think there is an easy way to test this now. I have tested this on my local, though.

Maybe we can return uuid value and assert it somehow, although its non-determenistic? I'm thinking if we can be protected from regression if anyone else changes this code later?

**Comment by huaxingao** [comment]
> Maybe we can return uuid value and assert it somehow, although its non-determenistic? I'm thinking if we can be protected from regression if anyone else changes this code later?

I am thinking of adding Iceberg tests in Comet after the Iceberg/Comet integration is complete, to ensure that the changes won't regress.

**Comment by huaxingao** [comment]
cc @viirya

File: common/src/main/java/org/apache/comet/parquet/ColumnReader.java [hunk 1]
```diff
@@ -169,6 +170,7 @@ public void close() {
 
   /** Returns a decoded {@link CometDecodedVector Comet vector}. */
   public CometDecodedVector loadVector() {
+
```
Comment (viirya) [inline common/src/main/java/org/apache/comet/parquet/ColumnReader.java:173]: Unnecessary change.

File: common/src/main/java/org/apache/comet/parquet/ColumnReader.java [hunk 2]
```diff
@@ -207,6 +214,7 @@ public CometDecodedVector loadVector() {
       DictionaryEncoding dictionaryEncoding = vector.getField().getDictionary();
 
       CometPlainVector cometVector = new CometPlainVector(vector, useDecimal128);
+      cometVector.setIsUuid(isUuid);
```
Comment (viirya) [inline common/src/main/java/org/apache/comet/parquet/ColumnReader.java:217]: Why not put it into constructor parameter list?

File: common/src/main/java/org/apache/comet/parquet/ColumnReader.java (hunk 1 above)
Comment (huaxingao) [reply common/src/main/java/org/apache/comet/parquet/ColumnReader.java]: remved extra line

File: common/src/main/java/org/apache/comet/parquet/ColumnReader.java (hunk 2 above)
Comment (huaxingao) [reply common/src/main/java/org/apache/comet/parquet/ColumnReader.java]: added `isUuid` in constructor parameter list

File: common/src/main/java/org/apache/comet/vector/CometPlainVector.java [hunk 3]
```diff
@@ -111,7 +115,12 @@ public UTF8String getUTF8String(int rowId) {
       byte[] result = new byte[length];
       Platform.copyMemory(
           null, valueBufferAddress + offset, result, Platform.BYTE_ARRAY_OFFSET, length);
-      return UTF8String.fromBytes(result);
+
+      if (!isUuid) {
+        return UTF8String.fromBytes(result);
+      } else {
+        return UTF8String.fromString(convertToUuid(result).toString());
+      }
```
Comment (viirya) [inline common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: For test, I think if it is possible to create a Parquet file with uuid column in unit test and read it? Due to this change the column should be read as uuid instead of string.

**Review by viirya — COMMENTED** [review]
The change looks okay. Just wondering if it is possible to add a test.

File: common/src/main/java/org/apache/comet/vector/CometPlainVector.java (hunk 3 above)
Comment (huaxingao) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: I looked `makeParquetFileAllTypes`, seems I can only use parquet type `INT64`, `INT32`, `BOOLEAN`, `BINARY`, `FLOAT`, `DOUBLE`, `INT96`, `FIXED_LEN_BYTE_ARRAY`. It doesn't seem I can use a UUID logical type in a Parquet schema. When I did my local test using iceberg, I was using iceberg's `UUIDType`

Comment (viirya) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: Hmm,

```
optional binary                   _13(ENUM);
```

I think we can specify physical type and logical type in the message, no? For example, above line specifies `binary` as physical type and `ENUM` as logical type.

Comment (viirya) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: So I guess you can try

```
optional binary                   _13(UUID);
```

?

Comment (huaxingao) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: I tried 
```
    val schemaStr =
      """
        |message root {
        |  required int32 id;
        |  required FIXED_LEN_BYTE_ARRAY(16) uuid (UUID);
        |}
      """.stripMargin
```
I got 
```
Fixed Binary size 36 does not match field type length 16
java.lang.IllegalArgumentException: Fixed Binary size 36 does not match field type length 16
	at org.apache.parquet.column.values.plain.FixedLenByteArrayPlainValuesWriter.writeBytes(FixedLenByteArrayPlainValuesWriter.java:56)
	at org.apache.parquet.column.impl.ColumnWriterBase.write(ColumnWriterBase.java:241)
	at org.apache.parquet.io.MessageColumnIO$MessageColumnIORecordConsumer.addBinary(MessageColumnIO.java:476)
	at org.apache.parquet.example.data.simple.BinaryValue.writeValue(BinaryValue.java:45)
	at org.apache.parquet.example.data.simple.SimpleGroup.writeValue(SimpleGroup.java:233)
	at org.apache.parquet.example.data.GroupWriter.writeGroup(GroupWriter.java:51)
	at org.apache.parquet.example.data.GroupWriter.write(GroupWriter.java:37)
	at org.apache.parquet.hadoop.example.GroupWriteSupport.write(GroupWriteSupport.java:85)
	at org.apache.parquet.hadoop.example.GroupWriteSupport.write(GroupWriteSupport.java:35)
	at org.apache.parquet.hadoop.InternalParquetRecordWriter.write(InternalParquetRecordWriter.java:138)
	at org.apache.parquet.hadoop.ParquetWriter.write(ParquetWriter.java:310)
	at org.apache.spark.sql.CometTestBase.$anonfun$createParquetFileWithIDAndUUID$1(CometTestBase.scala:482)
```

It has to be `FIXED_LEN_BYTE_ARRAY(16) uuid (UUID)`, otherwise I will get

```
UUID can only annotate FIXED_LEN_BYTE_ARRAY(16)
java.lang.IllegalStateException: UUID can only annotate FIXED_LEN_BYTE_ARRAY(16)
	at org.apache.parquet.Preconditions.checkState(Preconditions.java:196)
	at org.apache.parquet.schema.Types$BasePrimitiveBuilder$1.checkFixedPrimitiveType(Types.java:556)
```

Comment (huaxingao) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: Might be something wrong with the UUID value. I will check

Comment (huaxingao) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: I fixed the UUID data problem, but now I got `illegalParquetTypeError` from Spark. I don't think Spark supports Parquet's `UUID`. Iceberg maps `UUID` to `UTF8String`
```
Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0) (192.168.50.140 executor driver): org.apache.spark.sql.AnalysisException: Illegal Parquet type: FIXED_LEN_BYTE_ARRAY (UUID).
	at org.apache.spark.sql.errors.QueryCompilationErrors$.illegalParquetTypeError(QueryCompilationErrors.scala:1762)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.illegalType$1(ParquetSchemaConverter.scala:206)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.$anonfun$convertPrimitiveField$2(ParquetSchemaConverter.scala:310)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.convertPrimitiveField(ParquetSchemaConverter.scala:224)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.convertField(ParquetSchemaConverter.scala:187)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.$anonfun$convertInternal$3(ParquetSchemaConverter.scala:147)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.$anonfun$convertInternal$3$adapted(ParquetSchemaConverter.scala:117)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
	at scala.collection.immutable.Range.foreach(Range.scala:158)
	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.convertInternal(ParquetSchemaConverter.scala:117)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.convert(ParquetSchemaConverter.scala:87)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$.$anonfun$readSchemaFromFooter$2(ParquetFileFormat.scala:493)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$.readSchemaFromFooter(ParquetFileFormat.scala:493)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$.$anonfun$mergeSchemasInParallel$2(ParquetFileFormat.scala:473)
	at scala.collection.immutable.Stream.map(Stream.scala:418)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$.$anonfun$mergeSchemasInParallel$1(ParquetFileFormat.scala:473)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$.$anonfun$mergeSchemasInParallel$1$adapted(ParquetFileFormat.scala:464)
```

Comment (viirya) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: Hmm, okay, we still use Spark ParquetToSparkSchemaConverter to convert Parquet schema to Spark schema? I thought that we may have some custom one in Comet.

Comment (huaxingao) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: It failed [here](https://github.com/apache/datafusion-comet/blob/main/spark/src/test/scala/org/apache/spark/sql/CometTestBase.scala#L370). It's Spark code, not Comet code yet. That's why it uses Spark's `ParquetToSparkSchemaConverter`

Comment (viirya) [reply common/src/main/java/org/apache/comet/vector/CometPlainVector.java:123]: The failed one is Spark reads Parquet file using its data source. Can you try to read it with Comet? If Comet scan doesn't use `ParquetToSparkSchemaConverter`, I think we won't hit the above error?

File: common/src/main/java/org/apache/comet/parquet/CometParquetToSparkSchemaConverter.scala [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+          case _: BsonLogicalTypeAnnotation => BinaryType
+          case _: DecimalLogicalTypeAnnotation => makeDecimalType()
+          case _ => illegalType()
+        }
+
+      case FIXED_LEN_BYTE_ARRAY =>
+        typeAnnotation match {
+          case _: DecimalLogicalTypeAnnotation =>
+            makeDecimalType(Decimal.maxPrecisionForBytes(parquetType.getTypeLength))
+          case _: UUIDLogicalTypeAnnotation => StringType
```
Comment (huaxingao) [inline common/src/main/java/org/apache/comet/parquet/CometParquetToSparkSchemaConverter.scala:177]: All the code are copied from Spark's `ParquetToSparkSchemaConverter` except this line.

Comment (parthchandra) [reply common/src/main/java/org/apache/comet/parquet/CometParquetToSparkSchemaConverter.scala:259]: Are we adding this just to test? Or is this likely to be useful in other places?
Also, instead of copying, could we not just extend the Spark class and override `convertField`? WE can then call our impl of convertPrimitiveField for UUID and let the parent implementation handle the rest? 
We are likely to miss changes made in Spark if we make a copy.

Comment (huaxingao) [reply common/src/main/java/org/apache/comet/parquet/CometParquetToSparkSchemaConverter.scala:259]: The reason that I didn't override `convertField` is that it returns `ParquetColumn`, which is only exists in Spark3.3 and Spark3.4. Since we need to make Spark 3.2 work, so I made our own version of `CometParquetColumn` and was trying to make it work for Spark 3.2 too. It actually took quite some effort to make `CometParquetColumn` to work for all three version of Spark, because `ParquetColumn` uses different methods for Spark3.3 (which uses parquet 1.12.x) from Spark 3.4 (which uses parquet 1.13.x).

I had an offline discussion with @viirya. We will merge this PR without a test for now. After iceberg integration is done, we can probably add some iceberg test with uuid.

I will remove the tests for now.

**Comment by viirya** [comment]
Merged. Thanks @huaxingao @advancedxy @comphead @parthchandra

**Comment by huaxingao** [comment]
Thanks, everyone!
